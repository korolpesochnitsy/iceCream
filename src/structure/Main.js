import React, { useRef, useLayoutEffect } from 'react';
import { css } from 'emotion/macro';
import Helmet from 'react-helmet';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';

const mainStyle = css`
  max-width: 63.75em;
  margin-left: auto;
  margin-right: auto;
  min-height: 40em;
  padding-top: 2em;
  padding-bottom: 2em;
  outline: 0;

  .main-heading {
    font-family: 'cornerstone', sans-serif;
    padding: 1rem 0 2rem;
    color: #313030;
    font-size: 1.8em;
    outline: 0;
    text-align: center;
  }
`;

const Main = ({ headingText, headingLevel = 2, children, location }) => {
  const heading = useRef(null);
  const H = `h${headingLevel}`;

  useLayoutEffect(() => {
    if (location.state && location.state.focus) {
      heading.current.focus();
    }
    window.scrollTo(0, 0);
  }, [location.state]);

  return (
    <main className={mainStyle} tabIndex="-1" id="main">
      <Helmet>
        <title>{headingText} | Ice Cream</title>
      </Helmet>
      <H className="main-heading" ref={heading} tabIndex="-1">
        {headingText}
      </H>
      {children}
    </main>
  );
};

Main.propTypes = {
  headingText: PropTypes.string.isRequired,
  headingLevel: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  children: PropTypes.node.isRequired,
  location: PropTypes.shape({
    state: PropTypes.shape({
      focus: PropTypes.bool,
    }),
  }).isRequired,
};

export default withRouter(Main);
