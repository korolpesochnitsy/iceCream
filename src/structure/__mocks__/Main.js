import React from 'react';

const Main = ({ headingText, headingLevel = 2, children }) => {
  const H = `h${headingLevel}`;
  return (
    <main>
      <H>{headingText}</H>
      {children}
    </main>
  );
};

export default Main;
