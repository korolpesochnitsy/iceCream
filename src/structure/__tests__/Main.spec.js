import React from 'react';
import { render, cleanup } from '@testing-library/react';
import Helmet from 'react-helmet';
import Main from '../Main';

describe('Main', () => {
  let oldScrollTo = null;

  beforeEach(() => {
    oldScrollTo = global.window.scrollTo;
    global.window.scrollTo = jest.fn();
  });

  afterEach(() => {
    cleanup();
    global.window.scrollTo = oldScrollTo;
  });

  it('should render with standard heading and set the title', () => {
    const { container } = render(
      <Main headingText="Demo heading" location={{}}>
        <p>Page content</p>
      </Main>
    );

    expect(container.firstChild).toMatchSnapshot();
    expect(Helmet.peek().title).toEqual([
      'Demo heading',
      ' | Ultimate Ice Cream',
    ]);
    expect(window.scrollTo).toHaveBeenCalledWith(0, 0);
  });

  it('should focus the heading if the state dictates', () => {
    const { container } = render(
      <Main headingText="Demo heading" location={{ state: { focus: true } }}>
        <p>Page content</p>
      </Main>
    );

    const heading = container.firstChild.querySelector('h2');

    expect(document.activeElement).toEqual(heading);
  });

  it('should not focus the heading if the state is absent', () => {
    const { container } = render(
      <Main headingText="Demo heading" location={{}}>
        <p>Page content</p>
      </Main>
    );

    const heading = container.firstChild.querySelector('h2');

    expect(document.activeElement).not.toEqual(heading);
  });

  it('It should allow for custom heading levels', () => {
    const { container } = render(
      <Main headingText="Demo heading" headingLevel="3" location={{}}>
        <p>Page content</p>
      </Main>
    );

    expect(container.firstChild.querySelector('h3')).toHaveTextContent(
      'Demo heading'
    );
  });
});
