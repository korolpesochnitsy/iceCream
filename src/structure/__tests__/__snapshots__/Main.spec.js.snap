// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main should render with standard heading and set the title 1`] = `
.emotion-0 {
  max-width: 63.75em;
  margin-left: auto;
  margin-right: auto;
  min-height: 40em;
  padding-top: 2em;
  padding-bottom: 2em;
  outline: 0;
}

.emotion-0 .main-heading {
  font-family: 'cornerstone',sans-serif;
  padding: 1rem 0 2rem;
  color: #313030;
  font-size: 1.8em;
  outline: 0;
  text-align: center;
}

<main
  class="emotion-0"
  id="main"
  tabindex="-1"
>
  <h2
    class="main-heading"
    tabindex="-1"
  >
    Demo heading
  </h2>
  <p>
    Page content
  </p>
</main>
`;
